design/ceph-mon-pv.md

```diff
@@ -0,0 +1,139 @@
+# Ceph monitor PV storage
+
+Currently all of the storage for Ceph monitors (data, logs, etc..) is provided
+using HostPath volume mounts. Supporting PV storage for Ceph monitors in
+environments with dynamically provisioned network block devices (AWS, GCE,
+etc...) will allow monitors to migrate without requiring the monitor state to be
+rebuilt. This is particularly useful in cases where nodes fails or are taken
+offline for maintenance or upgrades.
+
+There are two major components to this design. First, the cluster CRD schema is
+modified to accept a PVC template for defining the storage to be used by
+monitors. And second, monitor pod management will be handled by StatefulSets
+rather than Deployments.
+
+# Monitor storage
+
+The monitor specification in the CRD is updated to include a persistent volume
+claim template that is used to obtain PVs for the monitor data store.
+
+```go
+type MonSpec struct {
+	Count                int  `json:"count"`
+	PreferredCount       int  `json:"preferredCount"`
+	AllowMultiplePerNode bool `json:"allowMultiplePerNode"`
+	VolumeClaimTemplate  *v1.PersistentVolumeClaim
+}
+```
+
+Unlike the storage requirements of OSDs (xf: StorageClassDeviceSets) it is
+reasonable to use default requirements for monitors (e.g. 5-10 GB). This means
+that a minimal claim may be specified using only the requested storage class in
+which case default resource requests will be used.
+
+```yaml
+volumeClaimTemplates:
+- spec:
+    storageClassName: gp2-ebs
+...
+volumeClaimTemplates:
+- spec:
+    storageClassName: local-storage
+```
+
+It is also reasonable for monitors to use storage in the form of a directory on
+the host's root file system. This behavior effectively provides something like
+Rook-specific dynamic HostPath provisioning and is the current default behavior
+in Rook. In order to continue to support this level of operational simplicity,
+the volume claim template itself can be left unspecified entirely to request the
+existing default behavior. Note that the storage class `local-storage` is not
+necessarily equivalent---there are a variety of implementations with the most
+popular requiring static provisioning [0, 1].
+
+[0]: https://github.com/rancher/local-path-provisioner
+[1]: https://github.com/kubernetes-sigs/sig-storage-local-static-provisioner
+
+## Non-PV monitor storage
+
+Not all monitor storage requirements should be satisfied using PV storage. Most
+notably log and crash data. These data may continue to be stored in HostPath
```

Comment:
IMHO, these first two sentences are not related to the "non-pv monitor storage" case but in addition. Even if you have monitor backed by PVs then `dataDirHostPath` does not go away for the reasons mentioned (logs and crashes). So I think we should move that outside or onto its own section.

Reply:
@leseb i'm pretty sure sure we are in agreement. "non-pv monitor storage" could potentially use a better wording---basically referring to the non-db storage used by a monitor. since we are effectively proposing that the db storage be moved to a pv, I didn't have a better phrase for htis.

@BlaineEXE i think that's a good question. i raise this at one point, and while I'm failing to remember the _exact_ line of reasoning, I'm pretty sure it had to do with log and crash data being unpredictable, so having it potentially compete with the monitor db storage both in terms of capacity and iops may not be ideal. i think there is another issue i'm forgetting.

Reply:
The goal is really to get away from using the host path. What if the cluster CR had another setting which is the PV template for logging?
- The daemon pods would each have their logging PVC
- If the pod ever fails over, the PVC would still be available where the new pod is started
- A scraping tool could `exec` into the pods to get the logs

The logging PVC would be a separate PR. For now I would suggest we allow for a pure PVC environment by skipping creating the logging hostpath if the dataDirHostPath is empty.